presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java

```diff
@@ -403,6 +403,72 @@ public static Slice upper(@SqlType("varchar(x)") Slice slice)
         return toUpperCase(slice);
     }
 
+    private static Slice pad(Slice text, long inputResultSize, Slice fillCharacters, int paddingOffset)
+    {
+        checkCondition(text.length() > 0, INVALID_FUNCTION_ARGUMENT, "Input string must not be empty");
+        checkCondition(inputResultSize > 0, INVALID_FUNCTION_ARGUMENT, "Result size must be greater than 0");
+        checkCondition(fillCharacters.length() > 0, INVALID_FUNCTION_ARGUMENT, "Fill character(s) must not be empty");
+
+        int textLength = countCodePoints(text);
+        int resultSize = Ints.saturatedCast(inputResultSize);
+
+        // if our result size is the same length as our string the return our string
+        if (textLength == resultSize) {
+            return text;
+        }
+
+        // if we're bigger as requested then truncate
+        if (textLength > resultSize) {
+            return SliceUtf8.substring(text, 0, resultSize);
+        }
+
+        // number of bytes in each code point and starting point of each code
+        // point of the fill characters
+        int fillCharactersLength = countCodePoints(fillCharacters);
+        int[] fillCharactersCounts = new int[fillCharactersLength];
+        for (int i = 0; i < fillCharactersLength; ++i) {
+          fillCharactersCounts[i] = lengthOfCodePointSafe(fillCharacters, offsetOfCodePoint(fillCharacters, i));
+        }
+
+        // figure out how many bytes we need to allocate
+        int numTextBytes = text.length();
+        int allocSize = numTextBytes;
+        for (int i = 0; i < resultSize - textLength; ++i) {
+          allocSize += fillCharactersCounts[i % fillCharactersLength];
+        }
+
+        Slice buffer = Slices.allocate(allocSize);
+
+        // fill in the existing string
+        int countBytes = allocSize - numTextBytes;
+        int startPointOfExistingText = (paddingOffset + countBytes) % allocSize;
+        buffer.setBytes(startPointOfExistingText, text);
+
+        for (int i = 0, index = 0, byteIndex = paddingOffset; i < resultSize - textLength; ++i, index = i % fillCharactersLength, byteIndex += fillCharactersCounts[index]) {
```

Comment:
Actually, given that we're pre-computing the size of the output buffer above, we should be able to blindly copy the bytes of the pad string until we need to handle the tail, and then just copy the portion that's needed to fit the output buffer.